File: developer/Samples/NET 4.6/WF_WCF_Samples/WCF/Basic/Binding/ByteStreamEncoder/VB/ByteStreamHttpBinding/ServiceModelEnumValidator.vb
'----------------------------------------------------------------
' Copyright (c) Microsoft Corporation.  All rights reserved.
'----------------------------------------------------------------

Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Configuration
Imports System.Reflection

Namespace Microsoft.Samples.ByteStreamEncoder

    Friend Class ServiceModelEnumValidator
        Inherits ConfigurationValidatorBase

        Private enumHelperType As Type
        Private isDefined As MethodInfo

        Public Sub New(ByVal enumHelperType As Type)
            Me.enumHelperType = enumHelperType
            Me.isDefined = Me.enumHelperType.GetMethod("IsDefined", BindingFlags.Static Or BindingFlags.NonPublic Or BindingFlags.Public)
        End Sub

        Public Overrides Function CanValidate(ByVal type As Type) As Boolean
            Return (Me.isDefined IsNot Nothing)
        End Function

        Public Overrides Sub Validate(ByVal value As Object)
            Dim retVal As Boolean = CBool(Me.isDefined.Invoke(Nothing, New Object() {value}))

            If (Not retVal) Then
                Dim isDefinedParameters() As ParameterInfo = Me.isDefined.GetParameters()
                Throw (New InvalidEnumArgumentException("value", CInt(Fix(value)), isDefinedParameters(0).ParameterType))
            End If
        End Sub
    End Class
End Namespace

File: developer/Samples/NET 4.6/WF_WCF_Samples/WCF/Basic/Binding/ByteStreamEncoder/CS/ByteStreamHttpBinding/ServiceModelEnumValidator.cs
//----------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
//----------------------------------------------------------------

using System;
using System.ComponentModel;
using System.Configuration;
using System.Reflection;

namespace Microsoft.Samples.ByteStreamEncoder
{

    class ServiceModelEnumValidator : ConfigurationValidatorBase
    {
        Type enumHelperType;
        MethodInfo isDefined;

        public ServiceModelEnumValidator(Type enumHelperType)
        {
            this.enumHelperType = enumHelperType;
            this.isDefined = this.enumHelperType.GetMethod("IsDefined", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
        }

        public override bool CanValidate(Type type)
        {
            return (this.isDefined != null);
        }

        public override void Validate(object value)
        {
            bool retVal = (bool)this.isDefined.Invoke(null, new object[] { value });

            if (!retVal)
            {
                ParameterInfo[] isDefinedParameters = this.isDefined.GetParameters();
                throw(new InvalidEnumArgumentException("value", (int)value, isDefinedParameters[0].ParameterType));
            }
        }
    }
}
